probes/internal/utils/permissions/permissions.go

```diff
@@ -0,0 +1,311 @@
+	probe string,
+	fs embed.FS,
+) (*finding.Finding, error) {
+	// Create finding
+	text, err := createText(r)
+	if err != nil {
+		return nil, fmt.Errorf("create finding: %w", err)
+	}
+	f, err := finding.NewWith(fs, probe,
+		text, nil, finding.OutcomeNegative)
+	if err != nil {
+		return nil, fmt.Errorf("create finding: %w", err)
+	}
+
+	// Create Location
+	var loc *finding.Location
+	if r.File != nil {
+		loc = &finding.Location{
+			Type:      r.File.Type,
+			Path:      r.File.Path,
+			LineStart: newUint(r.File.Offset),
+		}
+		if r.File.Snippet != "" {
+			loc.Snippet = newStr(r.File.Snippet)
+		}
+		f = f.WithLocation(loc)
+		f = f.WithRemediationMetadata(map[string]string{
+			"repo":     r.Remediation.Repo,
+			"branch":   r.Remediation.Branch,
+			"workflow": strings.TrimPrefix(f.Location.Path, ".github/workflows/"),
+		})
+	}
+	if r.LocationType != nil {
+		f = f.WithValue("permissionLocation", string(*r.LocationType))
+	}
+	if r.Name != nil {
+		f = f.WithValue("tokenName", *r.Name)
+	}
+	f = f.WithValue("permissionLevel", string(r.Type))
+	return f, nil
+}
+
+func CreateFindings(fs embed.FS,
+	raw *checker.RawResults,
+	locationType checker.PermissionLocation,
+	permissionLevel checker.PermissionLevel,
+	probe, tokenName string,
+) ([]finding.Finding, string, error) {
+	if raw == nil {
+		return nil, "", fmt.Errorf("%w: raw", uerror.ErrNil)
+	}
+
+	results := raw.TokenPermissionsResults
+	var findings []finding.Finding
+
+	if results.NumTokens == 0 {
+		f, err := finding.NewWith(fs, probe,
+			"No token permissions found",
+			nil, finding.OutcomeNotAvailable)
+		if err != nil {
+			return nil, probe, fmt.Errorf("create finding: %w", err)
+		}
+		findings = append(findings, *f)
+		return findings, probe, nil
+	}
+
+	for _, r := range results.TokenPermissions {
+		if r.Name == nil {
+			continue
+		}
+		if *r.Name != tokenName {
+			continue
+		}
+		if r.Type != permissionLevel {
+			continue
+		}
+		if *r.LocationType != locationType {
+			continue
+		}
+
+		// Create finding
+		f, err := CreateNegativeFinding(r, probe, fs)
+		if err != nil {
+			return nil, probe, fmt.Errorf("create finding: %w", err)
+		}
+		findings = append(findings, *f)
+	}
+
+	if len(findings) == 0 {
+		f, err := finding.NewWith(fs, probe,
+			fmt.Sprintf("no workflows with write permissions for '%v' at '%v'",
+				tokenName,
+				permissionLevel),
+			nil, finding.OutcomePositive)
+		if err != nil {
+			return nil, probe, fmt.Errorf("create finding: %w", err)
+		}
+		findings = append(findings, *f)
+	}
+	return findings, probe, nil
+}
+
+// avoid memory aliasing by returning a new copy.
+func newUint(u uint) *uint {
+	return &u
+}
+
+// avoid memory aliasing by returning a new copy.
+func newStr(s string) *string {
+	return &s
+}
+
+type TestData struct {
+	Name     string
+	Err      error
+	Raw      *checker.RawResults
+	Outcomes []finding.Outcome
+}
+
+func GetTests(locationType checker.PermissionLocation,
```

Comment:
should this live somewhere under `probes/internal/utils/test` instead?

Reply:
https://github.com/ossf/scorecard/pull/3816/commits/3d83a5d8b358a24aead499e223893ba17b787b16